src/runtime/current_thread/runtime.rs

```diff
@@ -409,4 +409,24 @@ impl Runtime {
             .try_with(|c| c.borrow().is_some())
             .unwrap_or(false)
     }
+
+    /// Enter the runtime context
+    pub fn enter<F, R>(&self, f: F) -> R
```

Comment:
So, this looks like the `enter` method on `tokio` 0.2's runtime, and that will work in some cases. However, `enter` requires us to be able to immutably borrow the runtime. This means that we cannot, for example, use `enter` on an external thread while we are running the runtime using `run` or `block_on` on a different thread.

I think we will also want to add an `enter` method to the `Handle` struct, like [`tokio` 0.2 has](https://docs.rs/tokio/0.2.9/tokio/runtime/struct.Handle.html#method.enter). That way, we can write code like this:
```rust
let rt = current_thread::Runtime::new().unwrap();
let handle = rt.handle().clone();
let thread1 = std::thread::spawn(move || {
    handle.enter(|| {
        futures::executor::block_on(async { // or whatever
            tokio::spawn(async {
                // ... whatever ...
            }).await;
            // ...
        });
    });
});

rt.block_on(async {
    // ...
});
thread1.join().unwrap();
```
This would probably necessitate some changes to the compat runtime's `Handle` type.

I think we could then implement `Runtime::enter` as
```rust
pub fn enter<F, R>(&self, f: F) -> R
where
    F: FnOnce() -> R,
{
    self.handle().enter(f)
}
```

Of course, this applies to the threadpool runtime as well.

Reply:
Absolutely agree. So we'd have `enter` on the threadpool's `TaskExecutor`, which is serving the purpose of a `Handle`.

I'll take a look at this in the next couple of days.

Reply:
For the threadpool, I would probably add a `Handle` type _in addition_ to the `TaskExecutor`, since `Handle` is an 0.2 API while `TaskExecutor` is an 0.1 API. I'd like to put the 0.2 methods (like `enter`) on a type with the same name as the 0.2 type (`Handle`), so that it's easier for users to transition from `tokio-compat` to using `tokio` 0.2 directly. Just my 2 cents.

Reply:
FWIW, I think we can add this stuff to the `Handle` types in a follow-up PR if you'd like to get this change merged sooner...